### ISSUE TITLE
`eigen` interface

### ISSUE THREAD

**Issue by mtfishman** [ITensor/ITensors.jl#331]
In `eigen`, in the default usage, it is easy to reproduce the original ITensor:
```julia
i = Index(2, "i")
j = Index(2, "j")
A = randomITensor(i, j, i', j')
A = A + swapprime(A, 0, 1)
U, D = eigen(A; ishermitian=true)
Ut = U'
B = U*D*dag(Ut)
@show norm(A - B)
```
However, in a more general use case, this isn't as easy:
```julia
i = Index(2, "i")
j = Index(2, "j")
Lis = IndexSet(i, j)
Ris = dag(sim(Lis))
A = randomITensor(Lis..., Ris...)
A = A + swapinds(A, Lis, Ris)  # Note: this function doesn't exist right now, we need to make it
U, D = eigen(A, Lis, Ris; ishermitian = true)
u = commonind(U, D)
# Annoying and error prone:
Ut = replaceinds(prime(U, u), Lis, Ris)
B = U * D * dag(Ut)
@show norm(A - B)
```
In general, we could return the ITensor `Ut` that has the proper indices such that `U*D*dag(Ut)` gives an approximation for the input (at least in the hermitian case). Returning `Ut` would only add the cost of the Index manipulations (while returning `dag(Ut)` would add the possible cost of the complex conjugation).

This came up for my when I was rewriting `_contract_densitymatrix`, and I found it useful to have the more general version of `eigen`, but I need both `U` and `Ut` and it was a bit annoying. The only downside I see is that it breaks code where someone was extracting the spectrum object from the `eigen` return, i.e.:
```julia
U,D,spec = eigen(A)
```
if we decide to make the return order `U,D,Ut,spec`. We could return `Ut` last, and people could extract it like:
```julia
F = eigen(A)
@show norm(F.U * F.D * F.Ut - A)
```
I'm not sure how common it will be to want `Ut` vs. the Spectrum object.

**Comment by mtfishman** [comment]
Just want to point out this is also useful in the non-Hermitian case, since `Ut` is the ITensor needed to test the eigenvectors:
```julia
i = Index(2, "i")
j = Index(2, "j")
Lis = IndexSet(i, j)
Ris = sim(Lis)
A = randomITensor(Lis..., Ris...)
U, D = eigen(A, Lis, Ris)
u = commonind(U, D)
# Annoying and error prone:
Ut = replaceinds(prime(U, u), Lis, Ris)
@show norm(U*D - D*dag(Ut))
```

@emstoudenmire, do you have an opinion on if/how we should return the `Ut` ITensor? We could even do something clever like return `Ud = dag(Ut)`, but only compute `dag` if someone asks for that return (for example in `TruncEigen`, store `Ut`, but then in the iterator compute `dag(Ut)`).

**Comment by emstoudenmire** [comment]
Hi Matt, yes I agree there's definitely some improvement to be done here. How about we discuss this in more detail later today via Zoom or on Slack? 

Tentatively, my current take on this is that the first case is special enough (yet also common) that it deserves its own function name, such as eigenh or eigen_hermitian. Even though Julia may just call various cases all "eigen", that's ok there because all matrices are rather similar so it makes more sense there. But here there's a rather different set of conventions and expectations about the Index structure of the input, and about the output. So forcing it to be the same function as a more general eigen could be the root of the problem here.

Then we could have a function that's just called "eigen" which doesn't assume anything about the ITensor or its index structure.

I'm open to other ideas, though, so let's discuss -

**Comment by mtfishman** [comment]
This is closed by #334.
